### PR TITLE
SI-7290 Discard duplicates in switchable alternative patterns.

### PR DESCRIPTION
**Pull request by retronym**
The pattern matcher must not allow duplicates to hit the
backend when generating switches. It already eliminates then
if they appear on different cases (with an unreachability warning.)

This commit does the same for duplicated literal patterns in an
alternative pattern: discard and warn.

Review by @adriaanm

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala [hunk 1]
```diff
@@ -22,7 +22,7 @@ import scala.reflect.internal.util.NoPosition
 trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
   import PatternMatchingStats._
   import global.{Tree, Type, Symbol, NoSymbol, CaseDef, atPos,
-    ConstantType, Literal, Constant, gen, EmptyTree,
+    ConstantType, Literal, Constant, gen, EmptyTree, distinctBy,
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala:25]: NICE USE OF distinctBy! You Might Be The First!

File: src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala [hunk 2]
```diff
@@ -452,7 +452,17 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
 
                 // succeed if they were all switchable
                 sequence(switchableAlts) map { switchableAlts =>
-                  CaseDef(Alternative(switchableAlts), guard, body)
+                  def extractConst(t: Tree) = t match {
+                    case Literal(const) => const
+                    case _              => t
+                  }
+                  // SI-7290 Discard duplicate alternatives that would crash the backend
+                  def distinctAlts = distinctBy(switchableAlts)(extractConst)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala]: why is this a def? always evaluated twice

File: src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala [hunk 3]
```diff
@@ -452,7 +452,17 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
...
+                  if (distinctAlts.size < switchableAlts.size) {
+                    val duplicated = switchableAlts.groupBy(extractConst).flatMap(_._2.drop(1))
+                    global.currentUnit.warning(pos, s"Pattern contains duplicate alternatives: ${duplicated.mkString(", ")}")
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala:463]: I'd only report distinct duplicates, if you know what I mean :neckbeard:

**Comment by retronym** [comment]
@adriaanm done

**Pull request merged**